File: Problems/CombinationSum/CombinationSum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
      List<List<Integer>> list = new ArrayList<>();
      Arrays.sort(candidates);
      backtrack(list, new ArrayList<>(), candidates, target, 0);
      return list;  
    }
    public void backtrack(List<List<Integer>> list, List<Integer> temp, int[] nums, int remain, int start) {
        if(remain < 0) return;
        else if (remain == 0) list.add(new ArrayList<>(temp));
        else {
            for(int i = start; i < nums.length; i ++) {
                temp.add(nums[i]);
                backtrack(list, temp, nums, remain - nums[i], i);
                temp.remove(temp.size() - 1);
            }
        }
    }
}

File: Problems/CombinationSum/CombinationSum.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    vector<vector<int>> res;
    void possibleCombinations(vector<int>& candidates, int target,vector<int>& temp,int idx){
        if(target == 0){
            res.push_back(temp);
            return;
        }
        if(target < 0 || idx>=candidates.size())
            return;
        temp.push_back(candidates[idx]);
        possibleCombinations(candidates,target-candidates[idx],temp,idx);
        temp.pop_back();
        possibleCombinations(candidates,target,temp,idx+1);
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> temp;
        possibleCombinations(candidates,target,temp,0);
        return res;
    }
};